How to close a client connection using the threading interface while stuck in ClientConnection.recv

Hello, I am using the threading client implementation. The problem I am facing is when I try to close the connection while stuck in the **_ClientConnection.recv_** method, the network connection goes down for a while and then gets back online. 

Sometimes, I get an exception; other times, it has to wait until the closing time out.

Another issue is while receiving messages with **_ClientConnection.recv_**, if the network goes down, it stops receiving messages. 


Here is the minimal code:

Steps to reproduce (eventually)
1- Run the code
2- Disable the computer network connection
3- wait for at least 60 seconds
4- stop the program with Ctrl+C (KeyboardInterrupt)
5- Output: Sometimes I get the expected behavior but other I get an exception like this:

```python
Traceback (most recent call last):
  File "/Users/jmrbcu/.pyenv/versions/totaltrack3/lib/python3.11/site-packages/websockets/sync/connection.py", line 199, in recv
    return self.recv_messages.get(timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jmrbcu/.pyenv/versions/totaltrack3/lib/python3.11/site-packages/websockets/sync/messages.py", line 102, in get
    raise EOFError("stream of frames ended")
EOFError: stream of frames ended

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/jmrbcu/.pyenv/versions/3.11.1/lib/python3.11/concurrent/futures/_base.py", line 340, in _invoke_callbacks
    callback(self)
  File "/Volumes/Development/totaltrack/totaltrack/core/telephony/ringcentral_recorder1.py", line 185, in <lambda>
    future.add_done_callback(lambda f: future.result())
                                       ^^^^^^^^^^^^^^^
  File "/Users/jmrbcu/.pyenv/versions/3.11.1/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/jmrbcu/.pyenv/versions/3.11.1/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/Users/jmrbcu/.pyenv/versions/3.11.1/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/Development/totaltrack/totaltrack/core/telephony/ringcentral_recorder1.py", line 167, in start
    message = ws.recv()
              ^^^^^^^^^
  File "/Users/jmrbcu/.pyenv/versions/totaltrack3/lib/python3.11/site-packages/websockets/sync/connection.py", line 201, in recv
    raise self.protocol.close_exc from self.recv_events_exc
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jmrbcu/.pyenv/versions/totaltrack3/lib/python3.11/site-packages/websockets/protocol.py", line 223, in close_exc
    assert self.state is CLOSED, "connection isn't closed yet"
           ^^^^^^^^^^^^^^^^^^^^
AssertionError: connection isn't closed yet
```

**Example Code:**

```python
from concurrent.futures import ThreadPoolExecutor, wait

fmt = "%(asctime)s [%(levelname)s] %(name)s.%(funcName)s:%(lineno)d %(message)s"
logging.basicConfig(level=logging.INFO, format=fmt)
stop = threading.Event()

def start():
    while True:
        try:
            message = ws.recv()
            logger.info("New message received: %s", message)
        except TimeoutError:
            logger.info("Timeout occurred")
        except websockets.ConnectionClosed as error:
            logger.info("Connection closed: %s", error)
            return

with ThreadPoolExecutor() as pool:
    ws = connect(api.get_websocket_endpoint())
    future = pool.submit(start)
    future.add_done_callback(lambda f: future.result())

    try:
        wait([future])
    except KeyboardInterrupt:
        print("Stopping")
        stop.set()
        ws.close()
        print("Stopped")
```

--- a/docs/project/changelog.rst
+++ b/docs/project/changelog.rst
@@ -75,6 +75,10 @@ Bug fixes
   start the connection handler anymore when ``process_request`` or
   ``process_response`` returns an HTTP response.
 
+* Fixed a bug in the :mod:`threading` implementation that could lead to
+  incorrect error reporting when closing a connection while
+  :meth:`~sync.connection.Connection.recv` is running.
+
 13.0.1
 ------
 

--- a/src/websockets/asyncio/connection.py
+++ b/src/websockets/asyncio/connection.py
@@ -274,6 +274,8 @@ class Connection(asyncio.Protocol):
         try:
             return await self.recv_messages.get(decode)
         except EOFError:
+            # Wait for the protocol state to be CLOSED before accessing close_exc.
+            await asyncio.shield(self.connection_lost_waiter)
             raise self.protocol.close_exc from self.recv_exc
         except ConcurrencyError:
             raise ConcurrencyError(
@@ -329,6 +331,8 @@ class Connection(asyncio.Protocol):
             async for frame in self.recv_messages.get_iter(decode):
                 yield frame
         except EOFError:
+            # Wait for the protocol state to be CLOSED before accessing close_exc.
+            await asyncio.shield(self.connection_lost_waiter)
             raise self.protocol.close_exc from self.recv_exc
         except ConcurrencyError:
             raise ConcurrencyError(
@@ -864,6 +868,7 @@ class Connection(asyncio.Protocol):
         # raise an exception.
         if raise_close_exc:
             self.close_transport()
+            # Wait for the protocol state to be CLOSED before accessing close_exc.
             await asyncio.shield(self.connection_lost_waiter)
             raise self.protocol.close_exc from original_exc
 
@@ -926,11 +931,14 @@ class Connection(asyncio.Protocol):
         self.transport = transport
 
     def connection_lost(self, exc: Exception | None) -> None:
-        self.protocol.receive_eof()  # receive_eof is idempotent
+        # Calling protocol.receive_eof() is safe because it's idempotent.
+        # This guarantees that the protocol state becomes CLOSED.
+        self.protocol.receive_eof()
+        assert self.protocol.state is CLOSED
+
+        self.set_recv_exc(exc)
 
         # Abort recv() and pending pings with a ConnectionClosed exception.
-        # Set recv_exc first to get proper exception reporting.
-        self.set_recv_exc(exc)
         self.recv_messages.close()
         self.abort_pings()
 

--- a/src/websockets/sync/connection.py
+++ b/src/websockets/sync/connection.py
@@ -206,6 +206,8 @@ class Connection:
         try:
             return self.recv_messages.get(timeout)
         except EOFError:
+            # Wait for the protocol state to be CLOSED before accessing close_exc.
+            self.recv_events_thread.join()
             raise self.protocol.close_exc from self.recv_exc
         except ConcurrencyError:
             raise ConcurrencyError(
@@ -240,6 +242,8 @@ class Connection:
             for frame in self.recv_messages.get_iter():
                 yield frame
         except EOFError:
+            # Wait for the protocol state to be CLOSED before accessing close_exc.
+            self.recv_events_thread.join()
             raise self.protocol.close_exc from self.recv_exc
         except ConcurrencyError:
             raise ConcurrencyError(
@@ -629,8 +633,6 @@ class Connection:
             self.logger.error("unexpected internal error", exc_info=True)
             with self.protocol_mutex:
                 self.set_recv_exc(exc)
-            # We don't know where we crashed. Force protocol state to CLOSED.
-            self.protocol.state = CLOSED
         finally:
             # This isn't expected to raise an exception.
             self.close_socket()
@@ -738,6 +740,7 @@ class Connection:
         # raise an exception.
         if raise_close_exc:
             self.close_socket()
+            # Wait for the protocol state to be CLOSED before accessing close_exc.
             self.recv_events_thread.join()
             raise self.protocol.close_exc from original_exc
 
@@ -788,4 +791,11 @@ class Connection:
         except OSError:
             pass  # socket is already closed
         self.socket.close()
+
+        # Calling protocol.receive_eof() is safe because it's idempotent.
+        # This guarantees that the protocol state becomes CLOSED.
+        self.protocol.receive_eof()
+        assert self.protocol.state is CLOSED
+
+        # Abort recv() with a ConnectionClosed exception.
         self.recv_messages.close()

